base_layer/core/src/transactions/aggregated_body.rs

```diff
@@ -193,6 +195,86 @@ impl AggregateBody {
         fee
     }
 
+    /// This function will check spent kernel rules like tx lock height etc
+    pub fn check_kernel_rules(&self, height: u64) -> Result<(), TransactionError> {
+        for kernel in self.kernels() {
+            if kernel.lock_height > height {
+                debug!(target: LOG_TARGET, "Kernel lock height was not reached: {}", kernel);
+                return Err(TransactionError::InvalidKernel);
+            }
+        }
+        Ok(())
+    }
+
+    /// Run through the outputs of the block and check that
+    /// 1. There is exactly ONE coinbase output
+    /// 1. The output's maturity is correctly set
+    /// 1. The amount is correct.
+    pub fn check_coinbase_output(
+        &self,
+        reward: MicroTari,
+        consensus_constants: &ConsensusConstants,
+        factories: &CryptoFactories,
+        height: u64,
+    ) -> Result<(), TransactionError>
+    {
+        let mut aggregated_body = AggregateBody::empty();
+        let mut coinbase_counter = 0; // there should be exactly 1 coinbase
+        for utxo in self.outputs() {
+            if utxo.features.flags.contains(OutputFlags::COINBASE_OUTPUT) {
+                coinbase_counter += 1;
+                if utxo.features.maturity < (height + consensus_constants.coinbase_lock_height()) {
+                    warn!(target: LOG_TARGET, "Coinbase found with maturity set too low");
+                    return Err(TransactionError::InvalidCoinbase);
+                }
+                aggregated_body.add_output(utxo.clone());
+            }
+        }
+        if coinbase_counter != 1 {
+            warn!(
+                target: LOG_TARGET,
+                "{} coinbases found in body. Only a single coinbase is permitted.", coinbase_counter,
+            );
+            return Err(TransactionError::InvalidCoinbase);
+        }
+
+        let mut coinbase_counter = 0; // there should be exactly 1 coinbase kernel as well
+        for kernel in self.kernels() {
+            if kernel.features.contains(KernelFeatures::COINBASE_KERNEL) {
+                coinbase_counter += 1;
+                aggregated_body.add_kernel(kernel.clone());
+            }
+        }
+        if coinbase_counter != 1 {
+            warn!(
+                target: LOG_TARGET,
+                "{} coinbase kernels found in body. Only a single coinbase kernel is permitted.", coinbase_counter,
+            );
+            return Err(TransactionError::InvalidCoinbase);
+        }
+        aggregated_body
```

Comment:
`validate_internal_consistency` is a very expensive call and it doesn't look to be saving any lines of code -- what with constructing a whole new AggregateBody --since you only really want to check the CB amount is correct. The signature check & Range proof check is going to be called again at some point when the full body is validated.

So just calculate and check Is CBUtxo.commitment == CBExcess + (Sum(fees) + reward).H? 
Then the checks above can be dramatically simplified too, e.g. something like

`counter = self.kernels().map_and_filter(|k| if k.features.contains(KernelFeatures::COINBASE_KERNEL) {1} else {0}).sum()`

Reply:
Done. I think. 
You still need to find and map out the coinbase utxo and kernel to use its values.